backend/glp/service.py

```
@@ -0,0 +1,800 @@
+        observed_map = (
+            snapshot.builder_points if origin == ORIGIN_BUILDER
+            else snapshot.community_points
+        )
+        observed = observed_map.get(user.id, 0)
+        account = GLPAccount.objects.select_for_update().filter(user=user).first()
+        consumed_field = 'consumed_bp' if origin == ORIGIN_BUILDER else 'consumed_cp'
+        consumed = getattr(account, consumed_field) if account else 0
+        if account is None or observed >= consumed:
+            raise GLPServiceError(
+                f'No decrease to resolve: observed {observed} >= consumed {consumed}.'
+            )
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)